### PR TITLE
OSAC-374 - Invoke Powershell - crossplatform

### PR DESCRIPTION
**Pull request by radunegulescu2**

### CODE REVIEW

File: Activities/Scripting/UiPath.Scripting.Activities.Design/DesignerMetadata.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+                nameof(InvokePowerShellCore<object>.PowerShellVariables), argumentDictionaryEditor);
+
+            builder.AddCustomAttributes(typeof(InvokePowerShellCore<>),
+                new DesignerAttribute(typeof(InvokePowerShellCoreDesigner)));
+
+            AddDefaultTypeArgumentAttribute(builder);
+
+            var hiddenActivityAttribute = new BrowsableAttribute(false);
+            builder.AddCustomAttributes(typeof(ExecutePowerShellCore), hiddenActivityAttribute);
+
```
Comment (nicolae-anghel) [inline Activities/Scripting/UiPath.Scripting.Activities.Design/DesignerMetadata.cs:39]: Please remove extra spaces

File: Activities/Scripting/UiPath.Scripting.Activities/PowerShell/ExecutePowerShellCore.cs [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+using System.Management.Automation.Runspaces;
+using System.Threading;
+using UiPath.Scripting.Activities.Properties;
+
+
+namespace UiPath.Scripting.Activities.PowerShell
+{
+    public sealed class ExecutePowerShellCore : AsyncCodeActivity
+    {
+        private const string PowerShellGlobalVariableNamePrefix = "Global.";
```
Comment (nicolae-anghel) [inline Activities/Scripting/UiPath.Scripting.Activities/PowerShell/ExecutePowerShellCore.cs:21]: Please add comment header to all of the methods and properties in this file

File: Activities/Scripting/UiPath.Scripting.Activities/PowerShell/ExecutePowerShellCore.cs [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+            {
+                DisposePipeline(pipeline);
+            }
+        }
+    }
+
+    // An instance of this object is returned to the runtime upon beginning
+    // execution.  This also keeps a pointer to a callback function, which
+    // it calls when it completes.
+    internal class PipelineInvokerAsyncResult : IAsyncResult
```
Comment (nicolae-anghel) [inline Activities/Scripting/UiPath.Scripting.Activities/PowerShell/ExecutePowerShellCore.cs]: Please move this class in a different file, so that the code will be more organized.

File: Activities/Scripting/UiPath.Scripting.Activities/PowerShell/InvokePowerShellCore.cs [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+using System.Activities;
+using System.Activities.Expressions;
+using System.Activities.Statements;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.ComponentModel;
+using System.Management.Automation;
+using UiPath.Scripting.Activities.Properties;
+
+namespace UiPath.Scripting.Activities.PowerShell
```
Comment (nicolae-anghel) [inline Activities/Scripting/UiPath.Scripting.Activities/PowerShell/InvokePowerShellCore.cs]: Please add header comments to all the classes and properties.

File: Activities/Scripting/UiPath.Scripting.Activities/PowerShell/InvokePowerShellCore.cs [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+{
+    internal class IPSHelper
```
Comment (nicolae-anghel) [inline Activities/Scripting/UiPath.Scripting.Activities/PowerShell/InvokePowerShellCore.cs]: Please move this class in a different file.

File: Activities/Scripting/UiPath.Scripting.Activities/PowerShell/InvokePowerShellCore.cs [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+                                    Exception =  new LambdaValue<Exception>(context => exception.Get(context))
+                                }
+                            }
+                        }
+                    }
+                }
+            };
+        }
+    }
+}
```
Comment (nicolae-anghel) [inline Activities/Scripting/UiPath.Scripting.Activities/PowerShell/InvokePowerShellCore.cs]: Please add new line here.

File: Activities/Scripting/UiPath.Scripting.Activities/Properties/SharedResources.cs [hunk 7]
```diff
@@ -0,0 +1,7 @@
+﻿
+namespace UiPath.Shared.Localization
+{
+    class SharedResources : UiPath.Scripting.Activities.Properties.Resources
+    {
+    }
+}
```
Comment (nicolae-anghel) [inline Activities/Scripting/UiPath.Scripting.Activities/Properties/SharedResources.cs]: Please add new line here

File: Activities/Scripting/UiPath.Scripting.Activities/Resources/ActivitiesMetadata.json [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+        }
+      ]
+    }
+  ],
+  "DefaultActivityColor": "#526069",
+  "DefaultActivityNameBackgroundColor": "#FFFFFF",
+  "DefaultActivityIconKey": null,
+  "AssemblyIconKey": null,
+  "ResourceManagerName": "UiPath.Scripting.Activities.Properties.UiPath.Scripting.Activities"
+}
```
Comment (nicolae-anghel) [inline Activities/Scripting/UiPath.Scripting.Activities/Resources/ActivitiesMetadata.json]: Please add new line here.

**Review by nicolae-anghel — CHANGES_REQUESTED** [review]

File: Activities/Directory.build.targets [hunk 9]
```diff
@@ -1,23 +1,36 @@
...
+		<PackageReference Update="System.Activities.ViewModels" Version="1.0.0-20220422.1"/>
+		<PackageReference Update="System.Management.Automation" Version="7.2.5" />
+		<PackageReference Include="Microsoft.PowerShell.Commands.Diagnostics" Version="7.2.5"/>
+		<PackageReference Include="Microsoft.PowerShell.ConsoleHost" Version="7.2.5" />
+		<PackageReference Include="Microsoft.PowerShell.Commands.Utility" Version="7.2.5" />
+		<PackageReference Include="Microsoft.PowerShell.Commands.Management" Version="7.2.5" />
+		<PackageReference Include="Microsoft.WSMan.Management" Version="7.2.5" />
+		<PackageReference Include="Microsoft.CodeAnalysis.VisualBasic" Version="4.0.1" />
+	</ItemGroup>
+	<ItemGroup Condition="$(TargetFramework.StartsWith('net5'))">
```
Comment (balansbogdan) [inline Activities/Directory.build.targets:9]: The Scripting package should support .NET6 only.

File: Activities/Scripting/UiPath.Scripting.Activities/PowerShell/ExecutePowerShellCore.cs [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+
+        [LocalizedDisplayName(nameof(Resources.Activity_InvokePowerShellCore_Property_IsScript_Name))]
+        [LocalizedDescription(nameof(Resources.Activity_InvokePowerShellCore_Property_IsScript_Description))]
+        public bool IsScript { get; set; }
+
+        [LocalizedDisplayName(nameof(Resources.Activity_InvokePowerShellCore_Property_PowerShellVariables_Name))]
+        [LocalizedDescription(nameof(Resources.Activity_InvokePowerShellCore_Property_PowerShellVariables_Description))]
+        public IDictionary<string, Argument> PowerShellVariables { get; set; }
+
+        private void AddDictionaryOfArguments<T>(CodeActivityMetadata metadata, IList<RuntimeArgument> arguments,
```
Comment (balansbogdan) [inline Activities/Scripting/UiPath.Scripting.Activities/PowerShell/ExecutePowerShellCore.cs]: Private methods should be placed at the bottom of the class.

**Review by balansbogdan — CHANGES_REQUESTED** [review]

File: Activities/Scripting/UiPath.Scripting.Activities/PowerShell/InvokePowerShellCore.cs (hunk 5 above)
Comment (alcxs) [reply Activities/Scripting/UiPath.Scripting.Activities/PowerShell/InvokePowerShellCore.cs]: also, add xml comment

File: Activities/Directory.build.targets (hunk 9 above)
Comment (alcxs) [reply Activities/Directory.build.targets:9]: This affects ALL packages and is left here for the ones that still target net5, like Database and others. This should not be changed, because the other packages will no longer build properly.

Comment (balansbogdan) [reply Activities/Directory.build.targets:9]: Got it. Thanks!

File: Activities/Scripting/UiPath.Scripting.Activities.Design/DesignerMetadata.cs [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+        public void Register()
+        {
+            AttributeTableBuilder builder = new AttributeTableBuilder();
+
+            builder.AddCustomAttributes(typeof(InvokePowerShellCore<>), new DesignerAttribute(typeof(InvokePowerShellCoreDesigner)));
+
+            MetadataStore.AddAttributeTable(builder.CreateTable());
+
+            builder.AddCustomAttributes(typeof(InvokePowerShellCore<>),
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities.Design/DesignerMetadata.cs]: group all AddCustomAttributes into one per line, and grouped together. declare all variables at the beginning and do the adding of attributes grouped, so it's easier to read

File: Activities/Scripting/UiPath.Scripting.Activities.Design/PowerShell/InvokePowerShellCoreDesigner.xaml.cs [hunk 12]
```diff
@@ -0,0 +1,15 @@
+﻿//----------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities.Design/PowerShell/InvokePowerShellCoreDesigner.xaml.cs]: why is this here? please remove

File: Activities/Scripting/UiPath.Scripting.Activities.Design/Properties/SharedResources.cs [hunk 13]
```diff
@@ -0,0 +1,7 @@
+﻿
+namespace UiPath.Shared.Localization
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities.Design/Properties/SharedResources.cs:2]: remove newline above

File: Activities/Scripting/UiPath.Scripting.Activities.Design/Themes/Icons.xaml [hunk 14]
```diff
@@ -0,0 +1,5 @@
+<ResourceDictionary xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml">
+
+
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities.Design/Themes/Icons.xaml]: leave a blank comment if this is kept for future icons, else remove it. it makes no sense to have it blank

File: Activities/Scripting/UiPath.Scripting.Activities/PowerShell/InvokePowerShellCore.cs [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+                }
+                metadata.Bind(argument, ra);
+                metadata.AddArgument(ra);
+
+                childParameters.Add(name, Argument.CreateReference(argument, name) as InArgument);
+            }
+        }
+    }
+
+    public sealed class InvokePowerShellCore<TResult> : Activity
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities/PowerShell/InvokePowerShellCore.cs]: separate file please

File: Activities/Scripting/UiPath.Scripting.Activities/Properties/SharedResources.cs (hunk 7 above)
Comment (alcxs) [reply Activities/Scripting/UiPath.Scripting.Activities/Properties/SharedResources.cs]: and remove the one above, the first line

File: Activities/Scripting/UiPath.Scripting.Activities/Resources/ActivitiesMetadata.json [hunk 16]
```diff
@@ -0,0 +1,26 @@
+{
+  "AssemblyRelativePath": "UiPath.Scripting.Activities.dll",
+  "Activities": [
+    {
+      "AssemblyQualifiedName": null,
+      "FullName": "UiPath.Scripting.Activities.InvokePowerShellCore",
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities/Resources/ActivitiesMetadata.json:6]: why was this file added? is there a requirement for studio web support? if so, then view models must be implemented as well (most likely will be the case, since XAML designers will no loger be used)

File: Activities/Directory.build.targets [hunk 17]
```diff
@@ -1,23 +1,36 @@
...
-	<PackageReference Update="UiPath.Platform" Version="22.4.1"  PrivateAssets="All" />
-  </ItemGroup>
-  <PropertyGroup>
-    <DefineConstants Condition="'$(TargetFramework)' == 'net5.0'">$(DefineConstants);NETSTANDARD</DefineConstants>
-  </PropertyGroup>
+	<ItemGroup Condition="$(TargetFramework.StartsWith('net6'))">
+		<PackageReference Update="Uipath.Workflow.Runtime" Version="6.0.0-20220401-03" />
+		<PackageReference Update="UiPath.Workflow" Version="6.0.0-20220401-03" />
+		<PackageReference Update="System.Activities.Core.Presentation" Version="6.0.0-20220506.1" />
+		<PackageReference Update="System.Activities.ViewModels" Version="1.0.0-20220422.1"/>
```
Comment (alcxs) [inline Activities/Directory.build.targets:6]: check what is the latest version and upgrade accordingly

**Review by alcxs — CHANGES_REQUESTED** [review]
see individual schanges

**Review by nicolae-anghel — APPROVED** [review]

File: Activities/Scripting/UiPath.Scripting.Activities.Design/Editors/EditorResources.xaml.cs [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+using res = UiPath.Scripting.Activities.Design.Properties.Resources;
+
+namespace UiPath.Scripting.Activities.Design
+{
+    /// <summary>
+    /// Resource Dictionary for Custom Editors
+    /// </summary>
+    public partial class EditorResources
+    {
+        private static ResourceDictionary s_resources;
```
Comment (balansbogdan) [inline Activities/Scripting/UiPath.Scripting.Activities.Design/Editors/EditorResources.xaml.cs]: This field should be renamed to _resources

**Review by balansbogdan — CHANGES_REQUESTED** [review]
In general:
- Make sure the private members are prefixed with _
- The private methods should be placed at the bottom of the class

File: Activities/Directory.build.targets [hunk 19]
```diff
@@ -1,23 +1,36 @@
...
+		<PackageReference Update="System.Activities.Core.Presentation" Version="5.0.0-20210724.1" />
+		<PackageReference Update="System.Activities.ViewModels" Version="1.0.0-20220307.1"/>
+	</ItemGroup>
+	<ItemGroup Condition="'$(TargetFramework)' == 'netcoreapp3.1'">
+		<PackageReference Update="System.Activities" Version="1.0.0-20201105-02" />
+		<PackageReference Update="UiPath.Workflow" Version="1.0.0-20201105-02" />
+		<PackageReference Update="System.Activities.Core.Presentation" Version="1.0.0-alpha003" />
+	</ItemGroup>
+	<ItemGroup>
+		<PackageReference Update="UiPath.Robot.Api" Version="22.4.1" PrivateAssets="All" />
```
Comment (alcxs) [inline Activities/Directory.build.targets:27]: remove versions here to be taken from directory.build.targets

File: Activities/Directory.build.targets [hunk 20]
```diff
@@ -1,23 +1,36 @@
...
+		<PackageReference Update="UiPath.Activities.Api" Version="22.4.1" PrivateAssets="All" />
```
Comment (alcxs) [inline Activities/Directory.build.targets:28]: remove versions here to be taken from directory.build.targets

File: Activities/Directory.build.targets [hunk 21]
```diff
@@ -1,23 +1,36 @@
...
+		<PackageReference Update="UiPath.Activities.Contracts" Version="22.4.1" PrivateAssets="All" />
```
Comment (alcxs) [inline Activities/Directory.build.targets:29]: remove versions here to be taken from directory.build.targets

File: Activities/Directory.build.targets [hunk 22]
```diff
@@ -1,23 +1,36 @@
...
+		<PackageReference Update="UiPath.Telemetry.Client" Version="1.3.18" PrivateAssets="All" />
```
Comment (alcxs) [inline Activities/Directory.build.targets:30]: remove versions here to be taken from directory.build.targets

File: Activities/Directory.build.targets [hunk 23]
```diff
@@ -1,23 +1,36 @@
...
+		<PackageReference Update="UiPath.Platform" Version="22.4.1"  PrivateAssets="All" />
```
Comment (alcxs) [inline Activities/Directory.build.targets:31]: remove versions here to be taken from directory.build.targets

File: Activities/Scripting/UiPath.Scripting.Activities.Design/DesignerMetadata.cs [hunk 24]
```diff
@@ -0,0 +1,49 @@
...
+using System.Activities.Presentation.PropertyEditing;
+using System.ComponentModel;
+using System.Linq;
+using System.Management.Automation;
+using UiPath.Scripting.Activities.Design.Properties;
+using UiPath.Scripting.Activities.PowerShell;
+
+namespace UiPath.Scripting.Activities.Design
+{
+    public class DesignerMetadata : IRegisterMetadata
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities.Design/DesignerMetadata.cs:12]: if this is taken from antoher repo, leave a comment above

File: Activities/Scripting/UiPath.Scripting.Activities.Design/DesignerMetadata.cs [hunk 25]
```diff
@@ -0,0 +1,49 @@
...
+    {
+        private void AddDefaultTypeArgumentAttribute(AttributeTableBuilder builder)
+        {
+            var presentationDll = AppDomain.CurrentDomain.GetAssemblies().First(a => a.GetName().Name == "System.Activities.Presentation");
+            Type attrType = presentationDll.GetType("System.Activities.Presentation.FeatureAttribute");
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities.Design/DesignerMetadata.cs]: extract this into constants

File: Activities/Scripting/UiPath.Scripting.Activities.Design/DesignerMetadata.cs [hunk 26]
```diff
@@ -0,0 +1,49 @@
...
+using UiPath.Scripting.Activities.Design.Properties;
+using UiPath.Scripting.Activities.PowerShell;
+
+namespace UiPath.Scripting.Activities.Design
+{
+    public class DesignerMetadata : IRegisterMetadata
+    {
+        private void AddDefaultTypeArgumentAttribute(AttributeTableBuilder builder)
+        {
+            var presentationDll = AppDomain.CurrentDomain.GetAssemblies().First(a => a.GetName().Name == "System.Activities.Presentation");
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities.Design/DesignerMetadata.cs]: extract this into constants

File: Activities/Scripting/UiPath.Scripting.Activities.Design/DesignerMetadata.cs [hunk 27]
```diff
@@ -0,0 +1,49 @@
...
+            Type attrType = presentationDll.GetType("System.Activities.Presentation.FeatureAttribute");
+            Type argType = presentationDll.GetType("System.Activities.Presentation.UpdatableGenericArgumentsFeature");
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities.Design/DesignerMetadata.cs]: extract this into constants

File: Activities/Scripting/UiPath.Scripting.Activities.Design/Editors/ArgumentDictionaryEditor.cs [hunk 28]
```diff
@@ -0,0 +1,59 @@
...
+using System.Activities.Presentation.Converters;
+using System.Activities.Presentation.Model;
+using System.Activities.Presentation.PropertyEditing;
+using System.Collections.Generic;
+using System.Linq;
+using System.Windows;
+
+namespace UiPath.Scripting.Activities.Design
+{
+    internal class ArgumentDictionaryEditor : DialogPropertyValueEditor
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities.Design/Editors/ArgumentDictionaryEditor.cs:15]: xml comment

File: Activities/Scripting/UiPath.Scripting.Activities.Design/Editors/PasswordBoxAssistant.cs [hunk 29]
```diff
@@ -0,0 +1,100 @@
+﻿#if !NETSTANDARD
+using System.Windows;
+using System.Windows.Controls;
+
+namespace UiPath.Scripting.Activities.Design
+{
+    public class PasswordBoxAssistant
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities.Design/Editors/PasswordBoxAssistant.cs:10]: xml comment

File: Activities/Scripting/UiPath.Scripting.Activities.Design/PowerShell/InvokePowerShellCoreDesigner.xaml [hunk 30]
```diff
@@ -0,0 +1,60 @@
+﻿<!-- Copyright (c) Microsoft Corporation.  All rights reserved. -->
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities.Design/PowerShell/InvokePowerShellCoreDesigner.xaml]: why is this here?

File: Activities/Scripting/UiPath.Scripting.Activities.Design/PowerShell/InvokePowerShellCoreDesigner.xaml.cs [hunk 31]
```diff
@@ -0,0 +1,11 @@
+﻿namespace UiPath.Scripting.Activities.Design
+{
+    // Interaction logic for InvokePowerShellCoreDesigner.xaml
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities.Design/PowerShell/InvokePowerShellCoreDesigner.xaml.cs]: xml comment not single line

File: Activities/Scripting/UiPath.Scripting.Activities.Design/Properties/SharedResources.cs (hunk 13 above)
Comment (alcxs) [reply Activities/Scripting/UiPath.Scripting.Activities.Design/Properties/SharedResources.cs:2]: still there :)

File: Activities/Scripting/UiPath.Scripting.Activities.Design/UiPath.Scripting.Activities.Design.csproj [hunk 32]
```diff
@@ -0,0 +1,56 @@
+﻿<Project Sdk="Microsoft.NET.Sdk.WindowsDesktop" ToolsVersion="Current">
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities.Design/UiPath.Scripting.Activities.Design.csproj]: `<Project Sdk="Microsoft.NET.Sdk">` only - see [docs](https://docs.microsoft.com/en-us/dotnet/core/project-sdk/overview)

File: Activities/Scripting/UiPath.Scripting.Activities/Localization.cs [hunk 33]
```diff
@@ -0,0 +1,55 @@
+﻿using System;
+using System.ComponentModel;
+using UiPath.Scripting.Activities.Properties;
+
+namespace UiPath.Scripting.Activities
+{
+    [AttributeUsage(AttributeTargets.Property)]
+    public class LocalizedCategoryAttribute : CategoryAttribute
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities/Localization.cs:11]: xml comment

File: Activities/Scripting/UiPath.Scripting.Activities/Localization.cs [hunk 34]
```diff
@@ -0,0 +1,55 @@
...
+        }
+
+        protected override string GetLocalizedString(string value)
+        {
+            return Resources.ResourceManager.GetString(value) ?? base.GetLocalizedString(value);
+        }
+    }
+
+    [AttributeUsage(AttributeTargets.Property | AttributeTargets.Class)]
+    public class LocalizedDisplayNameAttribute : DisplayNameAttribute
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities/Localization.cs:28]: xml comment

File: Activities/Scripting/UiPath.Scripting.Activities/Localization.cs [hunk 35]
```diff
@@ -0,0 +1,55 @@
...
+        public override string DisplayName
+        {
+            get
+            {
+                return Resources.ResourceManager.GetString(DisplayNameValue) ?? base.DisplayName;
+            }
+        }
+    }
+
+    public class LocalizedDescriptionAttribute : DescriptionAttribute
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities/Localization.cs:48]: xml comment

File: Activities/Scripting/UiPath.Scripting.Activities/PowerShell/ExecutePowerShellCore.cs [hunk 36]
```diff
@@ -0,0 +1,302 @@
...
+        public bool IsScript { get; set; }
+
+        /// <summary>
+        /// A dictionary of named objects that represent variables used within the current session of the command. A PowerShell command can retrieve information from IN and In/Out variables and can set Out variables.
+        /// </summary>
+        [LocalizedDisplayName(nameof(Resources.Activity_InvokePowerShellCore_Property_PowerShellVariables_Name))]
+        [LocalizedDescription(nameof(Resources.Activity_InvokePowerShellCore_Property_PowerShellVariables_Description))]
+        public IDictionary<string, Argument> PowerShellVariables { get; set; }
+
+        // Called before workflow execution to inform the runtime about arguments.
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities/PowerShell/ExecutePowerShellCore.cs]: xml comment, not single line

File: Activities/Scripting/UiPath.Scripting.Activities/PowerShell/ExecutePowerShellCore.cs [hunk 37]
```diff
@@ -0,0 +1,302 @@
...
+            // Overridding base.OnGetArguments to avoid reflection costs
+            RuntimeArgument inputArgument = new RuntimeArgument("Input", typeof(Collection<PSObject>), ArgumentDirection.In);
+            metadata.Bind(this.Input, inputArgument);
+            arguments.Add(inputArgument);
+
+            RuntimeArgument pipelineOutputArgument = new RuntimeArgument("PipelineOutput", typeof(Collection<PSObject>), ArgumentDirection.Out);
+            metadata.Bind(this.PipelineOutput, pipelineOutputArgument);
+            arguments.Add(pipelineOutputArgument);
+
+            RuntimeArgument errorsArgument = new RuntimeArgument("Errors", typeof(Collection<ErrorRecord>), ArgumentDirection.Out);
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities/PowerShell/ExecutePowerShellCore.cs]: extract magic string into constant

File: Activities/Scripting/UiPath.Scripting.Activities/PowerShell/ExecutePowerShellCore.cs [hunk 38]
```diff
@@ -0,0 +1,302 @@
...
+        protected override void CacheMetadata(CodeActivityMetadata metadata)
+        {
+            Collection<RuntimeArgument> arguments = new Collection<RuntimeArgument>();
+
+            // Overridding base.OnGetArguments to avoid reflection costs
+            RuntimeArgument inputArgument = new RuntimeArgument("Input", typeof(Collection<PSObject>), ArgumentDirection.In);
+            metadata.Bind(this.Input, inputArgument);
+            arguments.Add(inputArgument);
+
+            RuntimeArgument pipelineOutputArgument = new RuntimeArgument("PipelineOutput", typeof(Collection<PSObject>), ArgumentDirection.Out);
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities/PowerShell/ExecutePowerShellCore.cs]: extract magic string into constant

File: Activities/Scripting/UiPath.Scripting.Activities/PowerShell/ExecutePowerShellCore.cs [hunk 39]
```diff
@@ -0,0 +1,302 @@
...
+        [LocalizedDescription(nameof(Resources.Activity_InvokePowerShellCore_Property_PowerShellVariables_Description))]
+        public IDictionary<string, Argument> PowerShellVariables { get; set; }
+
+        // Called before workflow execution to inform the runtime about arguments.
+        protected override void CacheMetadata(CodeActivityMetadata metadata)
+        {
+            Collection<RuntimeArgument> arguments = new Collection<RuntimeArgument>();
+
+            // Overridding base.OnGetArguments to avoid reflection costs
+            RuntimeArgument inputArgument = new RuntimeArgument("Input", typeof(Collection<PSObject>), ArgumentDirection.In);
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities/PowerShell/ExecutePowerShellCore.cs]: extract magic string into constant

File: Activities/Scripting/UiPath.Scripting.Activities/PowerShell/ExecutePowerShellCore.cs [hunk 40]
```diff
@@ -0,0 +1,302 @@
...
+            }
+
+            if (this.PowerShellVariables.Count != 0)
+            {
+                //Prefix variable name so the name is unique from runtime's point of view,
+                //this is to handle the case where user might use the same name for variables and parameters
+                AddDictionaryOfArguments<Argument>(metadata, arguments, this.PowerShellVariables, PowerShellGlobalVariableNamePrefix, false);
+            }
+
+            arguments.Add(new RuntimeArgument("CommandText", typeof(String), ArgumentDirection.In));
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities/PowerShell/ExecutePowerShellCore.cs]: extract magic string into constant

File: Activities/Scripting/UiPath.Scripting.Activities/PowerShell/ExecutePowerShellCore.cs [hunk 41]
```diff
@@ -0,0 +1,302 @@
...
+            arguments.Add(new RuntimeArgument("ContinueOnError", typeof(bool), ArgumentDirection.In));
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities/PowerShell/ExecutePowerShellCore.cs]: extract magic string into constant

File: Activities/Scripting/UiPath.Scripting.Activities/PowerShell/ExecutePowerShellCore.cs [hunk 42]
```diff
@@ -0,0 +1,302 @@
...
+            Pipeline pipeline = context.UserState as Pipeline;
+            if (pipeline != null)
+            {
+                pipeline.Stop();
+                DisposePipeline(pipeline);
+            }
+            base.Cancel(context);
+        }
+
+        // Called by the runtime to begin execution of this asynchronous activity.
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities/PowerShell/ExecutePowerShellCore.cs]: xml comment not single line

File: Activities/Scripting/UiPath.Scripting.Activities/PowerShell/ExecutePowerShellCore.cs [hunk 43]
```diff
@@ -0,0 +1,302 @@
...
+                {
+                    throw;
+                }
+            }
+
+            context.UserState = pipeline;
+            return new PipelineInvokerAsyncResult(pipeline, callback, state);
+        }
+
+        // Called by the runtime after execution of this asynchronous activity.
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities/PowerShell/ExecutePowerShellCore.cs]: xml comment, not single line

File: Activities/Scripting/UiPath.Scripting.Activities/PowerShell/IPSHelper.cs [hunk 44]
```diff
@@ -0,0 +1,76 @@
+﻿using System.Activities;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Management.Automation;
+using UiPath.Scripting.Activities.Properties;
+
+namespace UiPath.Scripting.Activities.PowerShell
+{
+    public sealed class IPSHelper
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities/PowerShell/IPSHelper.cs:12]: xml comment

File: Activities/Scripting/UiPath.Scripting.Activities/PowerShell/PipelineInvokerAsyncResult.cs [hunk 45]
```diff
@@ -0,0 +1,187 @@
...
+using System.Collections.ObjectModel;
+using System.Management.Automation;
+using System.Management.Automation.Runspaces;
+using System.Threading;
+
+namespace UiPath.Scripting.Activities.PowerShell
+{
+    // An instance of this object is returned to the runtime upon beginning
+    // execution.  This also keeps a pointer to a callback function, which
+    // it calls when it completes.
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities/PowerShell/PipelineInvokerAsyncResult.cs]: xml comment, not like this

File: Activities/Scripting/UiPath.Scripting.Activities/UiPath.Scripting.Activities.csproj [hunk 46]
```diff
@@ -0,0 +1,46 @@
...
+	<PropertyGroup>
+		<PackageId>.noconflict</PackageId>
+		<TargetFrameworks>net6.0</TargetFrameworks>
+		<OutputPath>..\..\Output\Activities\Scripting\</OutputPath>
+	</PropertyGroup>
+	<ItemGroup Condition=" '$(TargetFramework)' == 'net6.0' ">
+		<PackageReference Include="UiPath.Workflow.Runtime" />
+		<PackageReference Include="UiPath.Workflow" />
+		<PackageReference Include="System.Activities.ViewModels" />
+		<PackageReference Include="System.Activities.Metadata" Version="0.1.0"/>
```
Comment (alcxs) [inline Activities/Scripting/UiPath.Scripting.Activities/UiPath.Scripting.Activities.csproj]: what is this version here? seems odd.

File: Activities/Shared/UiPath.Shared.Activities/Continu`ableAsyncCodeActivity.cs [hunk 47]
```diff
@@ -0,0 +1,56 @@
+﻿using System;
+using System.Activities;
+using System.Diagnostics;
+using System.Threading.Tasks;
+
+namespace UiPath.Shared.Activities
+{
+    public abstract class ContinuableAsyncCodeActivity : AsyncTaskCodeActivity
```
Comment (alcxs) [inline Activities/Shared/UiPath.Shared.Activities/Continu`ableAsyncCodeActivity.cs]: xml comment

**Review by alcxs — CHANGES_REQUESTED** [review]
see individual changes

**Comment by radunegulescu2** [comment]
Closed because it also needs integration with Studio Web.

**Comment by alexandru-petre** [comment]
No longer necessary due to recent updates of the InvokePowerShell activity from System.
